internal/services/cognitive/azure_ai_services_resource.go

```diff
@@ -0,0 +1,757 @@
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			client := metadata.Client.Cognitive.AccountsClient
+
+			var model AzureAIServicesResourceResourceModel
+
+			if err := metadata.Decode(&model); err != nil {
+				return err
+			}
+
+			id, err := cognitiveservicesaccounts.ParseAccountID(metadata.ResourceData.Id())
+			if err != nil {
+				return fmt.Errorf(" Cannot parse Azure AI service ID: %s", err)
+			}
+
+			networkACLs, subnetIds := expandAzureAIServicesNetworkACLs(model.NetworkACLs)
+			locks.MultipleByName(&subnetIds, network.VirtualNetworkResourceName)
+			defer locks.UnlockMultipleByName(&subnetIds, network.VirtualNetworkResourceName)
+
+			// also lock on the Virtual Network ID's since modifications in the networking stack are exclusive
+			virtualNetworkNames := make([]string, 0)
+			for _, v := range subnetIds {
+				subnetId, err := commonids.ParseSubnetIDInsensitively(v)
+				if err != nil {
+					return err
+				}
+				if !utils.SliceContainsValue(virtualNetworkNames, subnetId.VirtualNetworkName) {
+					virtualNetworkNames = append(virtualNetworkNames, subnetId.VirtualNetworkName)
+				}
+			}
+
+			locks.MultipleByName(&virtualNetworkNames, network.VirtualNetworkResourceName)
+			defer locks.UnlockMultipleByName(&virtualNetworkNames, network.VirtualNetworkResourceName)
+
+			props := cognitiveservicesaccounts.Account{
+				Sku: &cognitiveservicesaccounts.Sku{
+					Name: model.SkuName,
+				},
+				Properties: &cognitiveservicesaccounts.AccountProperties{
+					NetworkAcls:                   networkACLs,
+					CustomSubDomainName:           pointer.FromString(model.CustomSubdomainName),
+					AllowedFqdnList:               pointer.To(model.Fqdns),
+					PublicNetworkAccess:           pointer.To(cognitiveservicesaccounts.PublicNetworkAccess(model.PublicNetworkAccess)),
+					RestrictOutboundNetworkAccess: pointer.To(model.OutboundNetworkAccessRestricted),
+					DisableLocalAuth:              pointer.To(!model.LocalAuthorizationEnabled),
+					Encryption: &cognitiveservicesaccounts.Encryption{
+						KeySource: pointer.To(cognitiveservicesaccounts.KeySourceMicrosoftPointCognitiveServices),
+					},
+				},
+				Tags: pointer.To(model.Tags),
+			}
+
+			customMangedKey, err := expandAzureAIServicesCustomerManagedKey(model.CustomerManagedKey)
+			if err != nil {
+				return fmt.Errorf("expanding `customer_managed_key`: %+v", err)
+			}
+			if customMangedKey != nil {
+				props.Properties.Encryption = customMangedKey
+			}
+
+			expandIdentity, err := identity.ExpandSystemAndUserAssignedMapFromModel(model.Identity)
+			if err != nil {
+				return fmt.Errorf("expanding `identity`: %+v", err)
+			}
+			props.Identity = expandIdentity
+
+			future, err := client.AccountsUpdate(ctx, *id, props)
+			if err != nil {
+				return fmt.Errorf("updating %s: %+v", id, err)
+			}
+
+			if err := future.Poller.PollUntilDone(ctx); err != nil {
+				return fmt.Errorf("waiting for updating of %s: %+v", id, err)
+			}
```

Comment:
This is basically just a copy paste of the create function, please review [our guide on adding new resources](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/guide-new-resource.md), specifically the section on the update method and the use of `d.HasChange`.